feat(forms): add modules for forms and deprecatedForms

kara opened the pull request:
Currently, the deprecated form directives are still built-in platform directives, so they are provided by default.  We will want to remove them from default platform directives, so apps are not required to package forms with their app. This also makes forms friendly to offline compilation.

Instead, we'll expose three modules: 

OLD API:
- `DeprecatedFormsModule`

NEW API:
- `FormsModule`   (template-driven forms)
- `ReactiveFormsModule`  (reactive forms)

If you provide one of these modules, the default forms directives and providers from that module will be available to you app-wide.  Note: You can provide both the `FormsModule` and the `ReactiveFormsModule` together if you like, but they are fully-functional separately. 

**Before:**  (deprecated, but will still work in RC.5)

``` ts
import {disableDeprecatedForms, provideForms} from '@angular/forms';

bootstrap(App, [
   disableDeprecatedForms(),
   provideForms()
]);
```

**After:**

``` ts
import {DeprecatedFormsModule} from '@angular/common';

bootstrap(App, {modules: [DeprecatedFormsModule] });
```

-OR-

``` ts
import {FormsModule} from '@angular/forms';

bootstrap(App, {modules: [FormsModule] });
```

-OR-

``` ts
import {ReactiveFormsModule} from '@angular/forms';

bootstrap(App, {modules: [ReactiveFormsModule] });
```

You can also choose not to provide any forms module and run your app without forms.

r: @tbosch 
cc: @vsavkin

robwormald commented:
should this not have two modules (Reactive and Standard)?

tbosch commented on modules/@angular/forms/test/integration_spec.ts:
```diff
@@ -36,18 +34,15 @@ export function main() {
                 <input type="text" formControlName="login">
                </div>`;
 
-             tcb.overrideTemplate(MyComp8, t)
```
Could you change these tests as follows:

- have a `beforeEach( () => configureModule({modules: [FormsModule]}) );`
- remove the references to `FORMS_DIRECTIVES` / `FORMS_PROVIDES` from the test component, e.g. `MyComp8`

With that you test that the forms module actually works correctly.

kara replied:
Done!

tbosch commented:
Regarding reactive forms: 
Either add the `REACTIVE_FORM_DIRECTIVES` to the `FormsModule`, or create a `ReactiveFormsModule` that contains the `REACTIVE_FORM_DIRECTIVES` and the `FormBuilder` (and remove the `FormBuilder` from the `FormsModule`.

kara commented:
@robwormald Yeah, I had it functionally equivalent to what provideForms() was doing, but I think it makes more sense to have three modules, DeprecatedFormsModule, FormsModule, and ReactiveFormsModule.  Will update!

tbosch commented on modules/@angular/forms/src/directives.ts:
```diff
@@ -63,20 +74,12 @@ export {MaxLengthValidator, MinLengthValidator, PatternValidator, RequiredValida
+    /*@ts2dart_const*/[TEMPLATE_DRIVEN_DIRECTIVES, SHARED_FORM_DIRECTIVES];
 
-/**
+/**=
```
Extra `=`?

kara replied:
Good catch!  Removing.

tbosch commented:
I think this needs a `BREAKING CHANGE` as it removes the deprecated forms directives from bootstrap. you could note that we already announced this before.

kara committed:
feat(forms): add modules for forms and deprecatedForms

Closes #9732

BREAKING CHANGE:

We have removed the deprecated form directives from the built-in platform directive list, so apps are not required to package forms with their app. This also makes forms friendly to offline compilation.

Instead, we have exposed three modules:

OLD API:
- `DeprecatedFormsModule`

NEW API:
- `FormsModule`
- `ReactiveFormsModule`

If you provide one of these modules, the default forms directives and providers from that module will be available to you app-wide.  Note: You can provide both the `FormsModule` and the `ReactiveFormsModule` together if you like, but they are fully-functional separately.

**Before:**
```ts
import {disableDeprecatedForms, provideForms} from @angular/forms;

bootstrap(App, [
   disableDeprecatedForms(),
   provideForms()
]);
```

**After:**

```ts
import {DeprecatedFormsModule} from @angular/common;

bootstrap(App, {modules: [DeprecatedFormsModule] });
```

-OR-

```ts
import {FormsModule} from @angular/forms;

bootstrap(App, {modules: [FormsModule] });
```

-OR-

```ts
import {ReactiveFormsModule} from @angular/forms;

bootstrap(App, {modules: [ReactiveFormsModule] });
```

You can also choose not to provide any forms module and run your app without forms.

Or you can choose not to provide any forms module *and* provide form directives at will.  This will allow you to use the deprecatedForms API for some components and not others.

```
import {FORM_DIRECTIVES, FORM_PROVIDERS} from @angular/forms;

@Component({
   selector: some-comp,
   directives: [FORM_DIRECTIVES],
   providers: [FORM_PROVIDERS]
})
class SomeComp
```
